### DISCUSSION TITLE
Strawberry with federation2 and interface objects

### DISCUSSION THREAD

**Discussion by owl2** [strawberry-graphql/strawberry#3454] [Q&A] [answered]
Hello,

I have a question about this [https://strawberry.rocks/docs/guides/federation](tutorial): 

The federation2, as used in the documentation, is possible to request the books and add the reviews into the response, as resolvers.
But I wanted to know what to do if I want to request the reviews and pass the books as resolvers. In this tutorial, the graphql type functions can't solve my problem. But the interface object yes. I tried to implement this and I didn't suceed, because I have a 'resolve_type' function to implement. To be honest I'm not sure of what this function have to do. Can someone help me for this task ? 

My app books: 
```python
from typing import List
 
import strawberry
from strawberry.federation.schema_directives import Requires, External
 
 
# @strawberry.federation.type(keys=["id"])
# @strawberry.federation.interface(keys=["id"], directives=[External()])
@strawberry.federation.interface(keys=["id"],)
class Book:
    id: strawberry.ID
    title: str


    @classmethod
    def resolve_reference(cls, id: strawberry.ID, title: str) -> 'Book':
        # here we could fetch the book from the database
        # or even from an API
        return Book(id=id, title=title)
 
 
def get_all_books() -> List[Book]:
    return Book(id=strawberry.ID("1"), title="The Dark Tower")


 
@strawberry.type
class Query:
    all_books: List[Book] = strawberry.field(resolver=get_all_books)

    @strawberry.field
    def book(self) -> Book:
        return Book(id=strawberry.ID("1"), title='A title')
 
 
schema = strawberry.federation.Schema(
    query=Query,
    types=[Book], 
    enable_federation_2=True)
```

My app reviews: 
```python
from typing import List
 
import strawberry
from strawberry.federation.schema_directives import Requires, External
 
 
@strawberry.type
class Review:
    id: int
    body: str
    book: 'Book'
 
 
def get_reviews(root: "Book") -> List[Review]:
    return [
        Review(id=id_, body=f"A review for {root.id}")
        for id_ in range(root.reviews_count)
    ]
 
 
# @strawberry.federation.type(keys=["id"])
# @strawberry.federation.interface_object(keys=["id"], directives=[External()])
@strawberry.federation.interface_object(keys=["id"])
class Book:
    id: strawberry.ID
    reviews_count: int
    reviews: List[Review] = strawberry.field(resolver=get_reviews)
 
    @classmethod
    def resolve_reference(cls, id: strawberry.ID, reviews_count: int) -> 'Book':
        # here we could fetch the book from the database
        # or even from an API
        return Book(id=strawberry.ID("1"), reviews_count=3)


def get_all_reviews() -> List[Review]:
    return [Review(
                    id=strawberry.ID("1"), 
                    body="Blablabla", 
                    book=Book(id=strawberry.ID("1"), reviews_count=3),
                    )]

 
@strawberry.type
class Query:
    _hi: str = strawberry.field(resolver=lambda: "Hello World!")
    all_reviews: List[Review] = strawberry.field(resolver=get_all_reviews)
 
 
schema = strawberry.federation.Schema(
    query=Query, types=[Book, Review], enable_federation_2=True
)
```

The error I have: 
![image](https://github.com/strawberry-graphql/strawberry/assets/16677525/8ae4e0d8-c8c5-45d9-b498-c0654cda2119)

The federation is made with Apollo. 

Thank you :)

**Comment by patrick91** [comment]
@owl2 could you provide a repo for this? 😊

**Reply by owl2** [reply]
Hi @patrick91,
Here is the project I try to use, that is inspired from federation2 documentation.
[the project](https://github.com/owl2/gql-strawberry-interface-test)

I put the instructions in the README.md
The idea is to try to pass a request from the reviews with a books resolver. But I have an error with a simple query with the book entity.

Let me know if you want any additional information.

**Reply by patrick91** [reply, marked as answer]
Hi @owl2 thanks for the reproduction! I've made an example based on your repo that adds a `Reviewable` interface, which is used by the book and extended in the reviews service 😊

![CleanShot 2024-04-17 at 13 45 31@2x](https://github.com/strawberry-graphql/strawberry/assets/667029/1351ab73-f16e-4032-bc4a-2a3828861db9)


Here's the PR for it: https://github.com/owl2/gql-strawberry-interface-test/pull/1

**Reply by owl2** [reply]
Hi @patrick91 ,

Thank you very much for the answer. It helps me a lot.
